File: bindings/rust/idax/src/problem.rs
//! Typed analysis-problem lists.

use std::ffi::CString;

use crate::address::Address;
use crate::error::{self, Error, Result, Status};

/// A closed semantic analysis-problem category.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
#[repr(i32)]
pub enum Kind {
    MissingOffsetBase = 1,
    MissingName = 2,
    MissingForcedOperand = 3,
    MissingComment = 4,
    MissingReferences = 5,
    IgnoredJumpTable = 6,
    DisassemblyFailure = 7,
    AlreadyItemHead = 8,
    FlowBeyondLimits = 9,
    TooManyLines = 10,
    StackTraceFailure = 11,
    Attention = 12,
    AnalysisDecision = 13,
    RolledBackDecision = 14,
    FlairCollision = 15,
    FlairIndecision = 16,
}

/// Return a copied problem description, or `None` when none is recorded.
pub fn description(kind: Kind, address: Address) -> Result<Option<String>> {
    let mut out = std::ptr::null_mut();
    let status = unsafe { idax_sys::idax_problem_description(kind as i32, address, &mut out) };
    if status != 0 {
        return Err(error::consume_last_error("problem::description failed"));
    }
    if out.is_null() {
        return Ok(None);
    }
    unsafe {
        error::cstr_to_string_free(out, "problem::description returned invalid UTF-8").map(Some)
    }
}

/// Record a typed problem. `None` preserves the SDK default-message behavior.
pub fn remember(kind: Kind, address: Address, message: Option<&str>) -> Status {
    let message = message
        .map(|value| {
            CString::new(value)
                .map_err(|_| Error::validation("problem message contains an embedded NUL byte"))
        })
        .transpose()?;
    let pointer = message
        .as_ref()
        .map_or(std::ptr::null(), |value| value.as_ptr());
    let status = unsafe { idax_sys::idax_problem_remember(kind as i32, address, pointer) };
    error::int_to_status(status, "problem::remember failed")
}

/// Return the first problem address at or after the bound, or `None`.
pub fn next(kind: Kind, at_or_after: Address) -> Result<Option<Address>> {
    let mut out = 0;
    let mut has_value = 0;
    let status =
        unsafe { idax_sys::idax_problem_next(kind as i32, at_or_after, &mut out, &mut has_value) };
    if status != 0 {
        return Err(error::consume_last_error("problem::next failed"));
    }
    Ok((has_value != 0).then_some(out))
}

/// Remove a problem marker, returning whether it existed.
pub fn remove(kind: Kind, address: Address) -> Result<bool> {
    bool_result(
        |out| unsafe { idax_sys::idax_problem_remove(kind as i32, address, out) },
        "problem::remove failed",
    )
}

/// Return the copied short or long display name of a problem kind.
pub fn name(kind: Kind, long_form: bool) -> Result<String> {
    let mut out = std::ptr::null_mut();
    let status = unsafe { idax_sys::idax_problem_name(kind as i32, long_form as i32, &mut out) };
    if status != 0 {
        return Err(error::consume_last_error("problem::name failed"));
    }
    unsafe { error::cstr_to_string_free(out, "problem::name returned a null string") }
}

/// Return whether a typed problem exists at an address.
pub fn contains(kind: Kind, address: Address) -> Result<bool> {
    bool_result(
        |out| unsafe { idax_sys::idax_problem_contains(kind as i32, address, out) },
        "problem::contains failed",
    )
}

fn bool_result(function: impl FnOnce(*mut i32) -> i32, fallback: &str) -> Result<bool> {
    let mut out = 0;
    let status = function(&mut out);
    if status != 0 {
        return Err(error::consume_last_error(fallback));
    }
    Ok(out != 0)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn discriminants_match_the_pinned_sdk() {
        assert_eq!(Kind::MissingOffsetBase as i32, 1);
        assert_eq!(Kind::Attention as i32, 12);
        assert_eq!(Kind::FlairIndecision as i32, 16);
    }

    #[test]
    fn embedded_nul_is_rejected_before_ffi() {
        let error = remember(Kind::Attention, 0, Some("bad\0message")).unwrap_err();
        assert_eq!(error.category, crate::error::ErrorCategory::Validation);
    }

    #[test]
    fn signatures_preserve_optional_and_boolean_state() {
        let _: fn(Kind, Address) -> Result<Option<String>> = description;
        let _: fn(Kind, Address, Option<&str>) -> Status = remember;
        let _: fn(Kind, Address) -> Result<Option<Address>> = next;
        let _: fn(Kind, Address) -> Result<bool> = remove;
        let _: fn(Kind, bool) -> Result<String> = name;
        let _: fn(Kind, Address) -> Result<bool> = contains;
    }
}

File: bindings/swift/Sources/IDAX/Problem.swift
internal import CIDAX

/// A category in IDA's problem list.
///
/// Mirrors C++ `ida::problem::Kind`. The raw values are IDA's own and are not
/// contiguous with any other enumeration here.
public enum ProblemKind: Int32, Sendable {
    case missingOffsetBase = 1
    case missingName = 2
    case missingForcedOperand = 3
    case missingComment = 4
    case missingReferences = 5
    case ignoredJumpTable = 6
    case disassemblyFailure = 7
    case alreadyItemHead = 8
    case flowBeyondLimits = 9
    case tooManyLines = 10
    case stackTraceFailure = 11
    case attention = 12
    case analysisDecision = 13
    case rolledBackDecision = 14
    case flairCollision = 15
    case flairIndecision = 16
}

/// IDA's problem list — the addresses the analyser flagged as needing attention.
public enum Problem {

    /// IDA's description of the problem recorded at `address`.
    public static func description(
        of kind: ProblemKind,
        at address: Address
    ) throws(IDAError) -> String {
        var out: UnsafeMutablePointer<CChar>? = nil
        try checkStatus(
            idax_problem_description(kind.rawValue, address, &out),
            "problem.description"
        )
        return takeCString(out)
    }

    /// Records a problem of `kind` at `address`.
    public static func remember(
        _ kind: ProblemKind,
        at address: Address,
        message: String = ""
    ) throws(IDAError) {
        try checkStatus(
            idax_problem_remember(kind.rawValue, address, message),
            "problem.remember"
        )
    }

    /// The first address at or after `address` carrying a problem of `kind`.
    ///
    /// - Returns: `nil` when there is no such address, which is not an error.
    public static func next(
        _ kind: ProblemKind,
        atOrAfter address: Address
    ) throws(IDAError) -> Address? {
        var out: UInt64 = 0
        var hasValue: Int32 = 0
        try checkStatus(
            idax_problem_next(kind.rawValue, address, &out, &hasValue),
            "problem.next"
        )
        return hasValue != 0 ? out : nil
    }

    /// Removes the problem of `kind` recorded at `address`.
    ///
    /// - Returns: whether a problem was actually removed.
    @discardableResult
    public static func remove(
        _ kind: ProblemKind,
        at address: Address
    ) throws(IDAError) -> Bool {
        try withOutput("problem.remove", Int32(0)) {
            idax_problem_remove(kind.rawValue, address, $0)
        } != 0
    }

    /// IDA's name for a problem category.
    ///
    /// - Parameter longForm: request the descriptive sentence rather than the
    ///   short identifier.
    public static func name(
        of kind: ProblemKind,
        longForm: Bool = false
    ) throws(IDAError) -> String {
        var out: UnsafeMutablePointer<CChar>? = nil
        try checkStatus(
            idax_problem_name(kind.rawValue, longForm ? 1 : 0, &out),
            "problem.name"
        )
        return takeCString(out)
    }

    /// Whether a problem of `kind` is recorded at `address`.
    public static func contains(
        _ kind: ProblemKind,
        at address: Address
    ) throws(IDAError) -> Bool {
        try withOutput("problem.contains", Int32(0)) {
            idax_problem_contains(kind.rawValue, address, $0)
        } != 0
    }
}
